nwbwidgets/timeseries.py

```diff
@@ -967,3 +972,349 @@ def update(
         fig.data = []
         fig.layout = {}
         return self.plot_group(group_inds,data,time_ts_aligned,fig,order)
+
+
+def trialize_time_series(
+    time_series: TimeSeries, trials_table: Union[DynamicTable, pd.DataFrame], index: int = 0
+) -> pd.DataFrame:
+
+    if isinstance(trials_table, DynamicTable):
+        trials_table_df = trials_table.to_dataframe()
+    else:
+        trials_table_df = trials_table
+
+    if time_series.timestamps:
+        timestamps = time_series.timestamps[:]
+    else:
+        number_of_frames = time_series.data.shape[0]
+        dt = 1.0 / time_series.rate
+        time_offset = time_series.starting_time if time_series.starting_time else 0
+        timestamps = np.arange(0, number_of_frames) * dt + time_offset
+
+    trials_table_df = trials_table_df.reset_index().rename(columns={"id": "trial"})
+
+    # Map timestamps to the trial interval (start_time, stop_time)
+    start_times = trials_table_df.start_time.to_numpy()
+    stop_times = trials_table_df.stop_time.to_numpy()
+    larger_than_start_time = start_times[np.newaxis, :] < timestamps[:, np.newaxis]
+    smaller_than_stop_time = timestamps[:, np.newaxis] <= stop_times[np.newaxis, :]
+    timestamps_contained_in_trial = np.logical_and(larger_than_start_time, smaller_than_stop_time)
+    # First dimension is timestamp index second dimension is trial table index
+    data_index_within_trials, index_in_trial_table = np.nonzero(timestamps_contained_in_trial)
+
+    # Check efficiency of accessing HDF5 with indexes
+    data = time_series.data[data_index_within_trials, index]
+    data_df = pd.DataFrame(dict(timestamps=timestamps[data_index_within_trials], data=data))
+
+    # Create an start_time in the data table and merge withe the trial table using a left join operation
+    data_df["start_time"] = [trials_table_df.start_time[index] for index in index_in_trial_table]
+    data_df_trialized = pd.merge(left=data_df, right=trials_table_df, on="start_time", how="left")
+    data_df_trialized["centered_timestamps"] = data_df_trialized.timestamps - data_df_trialized.start_time
```

Comment:
can you use https://github.com/NeurodataWithoutBorders/nwbwidgets/blob/master/nwbwidgets/utils/timeseries.py#L244?

Reply:
Most of what that function is doing is to prepare a dataframe so the methods of ploly express can be used out of the box (otherwise faceting is way more complicated). The function you point out seems to return a list with the data form the trials, maybe it can indeed substitute a part of it.

Reply:
I ended up adding a vectorized version of the function that bisects in the utils.